[WebDriver] Add BiDi message handling support

lauromoura opened the pull request:
#### fb7cd496792a8bdb747212b73b53dc8acf8c3faf
<pre>
[WebDriver] Add BiDi message handling support
<a href="https://bugs.webkit.org/show_bug.cgi?id=280501">https://bugs.webkit.org/show_bug.cgi?id=280501</a>

Reviewed by Carlos Garcia Campos.

This commit adds the basic incoming message handling and replying. It
uses the simple `session.status` method.

As we support a single session, calling `session.status()` will always
return false for now, given we do not support calling static methods
without a session yet.

This version matches the incoming message with the message handler
only by the &quot;method&quot; name parameter, leaving the matching of the
method parameters to the handler function.

This commit also changes the Message data type to hold a CString instead
of pair of plain C string + length.

* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::handleMessage): Added.
(WebDriver::WebDriverService::findBidiCommand): Added.
(WebDriver::WebDriverService::bidiSessionStatus): Added.
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketMessageHandler::Message::fail): Use int for command
ids.
(WebDriver::WebSocketMessageHandler::Message::reply): Added.
* Source/WebDriver/WebSocketServer.h:
* Source/WebDriver/soup/WebSocketServerSoup.cpp:
(WebDriver::WebSocketServer::sendMessage):

amend: ensure we call the completion handler
Canonical link: <a href="https://commits.webkit.org/286076@main">https://commits.webkit.org/286076@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/495baddc3b5abf044ad659d07256579479ebf98d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74458 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27269 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78843 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25696 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1672 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58503 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16823 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77525 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48668 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64023 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38914 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45687 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21519 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24029 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67067 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21865 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80360 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1775 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1033 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66794 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1923 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64041 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66077 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16454 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10013 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8168 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1739 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1768 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1756 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1775 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,100 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+
+    if (!message.connection) {
+        WTFLogAlways("Incoming message without attached connection. Ignoring message.");
+        return;
```
I think we should always call the completion handler before returning, in this case generating an error

lauromoura replied:
Makes sense.

FTR, in the case of this specific error, it'll still be an early return on the soup server, as the server won't know which connection to send this reply to, relying on the original message's connection. But at least the original handler will be unblocked.

In the following check right after this one, I'll update it to send InvalidSessionID instead of ignoring the `completionHandler` too.

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,100 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+    result->setString("method"_s, method);
+    result->setObject("params"_s, *asObject->getObject("params"_s));
+    *handler = candidate->handler;
+    return result;
```
Why do we return a JSON object? I don't see where method is used, I guess we could make this bool and return id, handler and params as output parameters, like findCommand

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,100 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+        ((*this).*handler)(id, WTFMove(parameters), [completionHandler = WTFMove(completionHandler), message](std::optional<WebSocketMessageHandler::Message> resultMessage) {
+
+            // 6.7.5 If method is "session.new", let session be the entry in the list of active sessions whose session ID is equal to the "sessionId" property of value, append connection to session’s session WebSocket connections, and remove connection from the WebSocket connections not associated with a session.
+            // TODO https://bugs.webkit.org/show_bug.cgi?id=281722
```
```suggestion
            // FIXME: https://bugs.webkit.org/show_bug.cgi?id=281722
```

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,100 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+        WTFLogAlways("Failed to find appropriate BiDi command");
+        std::optional<int> commandId = std::nullopt;
+        auto parsedMessageObject = parsedMessageValue->asObject();
+        if (parsedMessageObject) {
```
```suggestion
        if (auto parsedMessageObject = parsedMessageValue->asObject()) {
```

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,100 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+        });
+    } else {
+        WTFLogAlways("Failed to find appropriate BiDi command");
+        std::optional<int> commandId = std::nullopt;
```
```suggestion
        std::optional<int> commandId;
```

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2579,6 +2674,18 @@ void WebDriverService::takeElementScreenshot(RefPtr<JSON::Object>&& parameters,
 }
 
 #if ENABLE(WEBDRIVER_BIDI)
+void WebDriverService::bidiSessionStatus(unsigned id, RefPtr<JSON::Object>&&, Function<void (std::optional<WebSocketMessageHandler::Message>)>&& completionHandler)
```
```suggestion
void WebDriverService::bidiSessionStatus(unsigned id, RefPtr<JSON::Object>&&, Function<void(std::optional<WebSocketMessageHandler::Message>)>&& completionHandler)
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -176,6 +176,24 @@ WebSocketMessageHandler::Message WebSocketMessageHandler::Message::fail(CommandR
+    reply->setInteger("id"_s, id);
+    auto resultObject = result->asObject();
+
+    if (resultObject)
```
```suggestion
    if (auto resultObject = result->asObject())
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -176,6 +176,24 @@ WebSocketMessageHandler::Message WebSocketMessageHandler::Message::fail(CommandR
+    else
+        reply->setObject("result"_s, JSON::Object::create());
+
+    auto serializedReply = reply->toJSONString().utf8().data();
```
This is the data of a temporary CString

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -176,6 +176,24 @@ WebSocketMessageHandler::Message WebSocketMessageHandler::Message::fail(CommandR
+        reply->setObject("result"_s, JSON::Object::create());
+
+    auto serializedReply = reply->toJSONString().utf8().data();
+    auto serializedReplyLength = strlen(serializedReply);
```
No need to strlen, just get the size of the utf8 CString

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -176,6 +176,24 @@ WebSocketMessageHandler::Message WebSocketMessageHandler::Message::fail(CommandR
+    auto serializedReply = reply->toJSONString().utf8().data();
+    auto serializedReplyLength = strlen(serializedReply);
+    // Connection will be set when sending the message back to the client, from the incoming message.
+    return { nullptr, serializedReply, serializedReplyLength };
```
The data is not copied here, so even if you avoid the temporary issue, the CString will be released when this function finishes. I guess WebSocketMessageHandler::Message should take ownership of the data, so instead of data + dataLength, it should have a CString

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -162,6 +162,17 @@ std::optional<String> WebSocketServer::listen(const String& host, unsigned port)
+{
+    for (const auto& pair : m_connectionToSession) {
+        if (pair.value == session) {
+            GRefPtr<GBytes> rawMessage = adoptGRef(g_bytes_new(message.utf8().data(), message.length()));
```
The lengght of the utf16 string might not be the same as the utf8 conversion, no? Why don't we just pass a CString to sendMessage? I guess all ports need the message as utf8

carlosgcampos replied:
Did you miss this comment?

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -176,6 +176,24 @@ WebSocketMessageHandler::Message WebSocketMessageHandler::Message::fail(CommandR
     return { *connection, serializedReply, strlen(serializedReply) };
```
I didn't noticed this in previous reviews, but this is wrong too.

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2579,6 +2674,18 @@ void WebDriverService::takeElementScreenshot(RefPtr<JSON::Object>&& parameters,
+    else
+        result->setString("message"_s, "Maximum number of sessions created"_s);
+
+    completionHandler(WebSocketMessageHandler::Message::reply("success"_s, id, result));
```
```suggestion
    completionHandler(WebSocketMessageHandler::Message::reply("success"_s, id, WTFMove(result)));
```

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,100 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+    if (candidate == std::end(s_bidiCommands))
+        return false;
+
+    auto result = JSON::Object::create();
```
This is unused

carlosgcampos commented on Source/WebDriver/WebSocketServer.h:
```diff
@@ -91,10 +91,10 @@ class WebSocketMessageHandler {
         Connection connection;
-        const char* data { nullptr };
-        size_t dataLength { 0 };
+        const CString payload;
```
This is not const, no?

lauromoura replied:
As we never modify the string after the message construction (only accessing the c string when sending back the reply), I went with const for it. Is there any reason to make it non-const?

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -92,21 +92,20 @@ static void handleWebSocketMessage(SoupWebsocketConnection* connection, SoupWebs
-    gconstpointer messageData = g_bytes_get_data(message, &messageSize);
-    WebSocketMessageHandler::Message messageObj = { connection, static_cast<const char*>(messageData), messageSize };
+    gconstpointer messageData = g_bytes_get_data(message, nullptr);
+    WebSocketMessageHandler::Message messageObj = { connection, static_cast<const char*>(messageData) };
```
Get the GBytes length too and build the CString with both data and length

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,99 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
 }
 
-void WebDriverService::handleMessage(WebSocketMessageHandler::Message&&, Function<void (WebSocketMessageHandler::Message&&)>&&)
+void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message, Function<void (WebSocketMessageHandler::Message&&)>&& completionHandler)
```
```suggestion
void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message, Function<void(WebSocketMessageHandler::Message&&)>&& completionHandler)
```

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,99 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+        return;
+    }
+
+    ((*this).*handler)(id, WTFMove(parameters), [completionHandler = WTFMove(completionHandler), message](std::optional<WebSocketMessageHandler::Message> resultMessage) {
```
Why is resultMessage optional? Don't we always have to send a response?

lauromoura replied:
Indeed. The WebDriver classic handlers return a `CommandResult&&`. I've updated the WebDriver BiDi ones to also move a message back to the caller.

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,99 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+        if (resultMessage) {
+            resultMessage->connection = message.connection;
+            completionHandler(WTFMove(*resultMessage));
+        }
```
Same question here, the completion handler should always be called, no?

lauromoura replied:
Ditto

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -407,11 +413,99 @@ bool WebDriverService::acceptHandshake(HTTPRequestHandler::Request&& request)
+            completionHandler(WTFMove(*resultMessage));
+        }
+    });
+
```
```suggestion
```

lauromoura committed:
[WebDriver] Add BiDi message handling support
https://bugs.webkit.org/show_bug.cgi?id=280501

Reviewed by Carlos Garcia Campos.

This commit adds the basic incoming message handling and replying. It
uses the simple `session.status` method.

As we support a single session, calling `session.status()` will always
return false for now, given we do not support calling static methods
without a session yet.

This version matches the incoming message with the message handler
only by the "method" name parameter, leaving the matching of the
method parameters to the handler function.

This commit also changes the Message data type to hold a CString instead
of pair of plain C string + length.

* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::handleMessage): Added.
(WebDriver::WebDriverService::findBidiCommand): Added.
(WebDriver::WebDriverService::bidiSessionStatus): Added.
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketMessageHandler::Message::fail): Use int for command
ids.
(WebDriver::WebSocketMessageHandler::Message::reply): Added.
* Source/WebDriver/WebSocketServer.h:
* Source/WebDriver/soup/WebSocketServerSoup.cpp:
(WebDriver::WebSocketServer::sendMessage):

amend: ensure we call the completion handler
Canonical link: https://commits.webkit.org/286076@main
